packages/hub/src/graphql/queries/search.ts

```diff
@@ -0,0 +1,126 @@
+      resolve: (obj) => {
+        const snippet = (obj.node as SearchableWithEdgeData).slugSnippet;
+        return String(snippet ?? "");
+      },
+    }),
+    textSnippet: t.string({
+      resolve: (obj) => {
+        const snippet = (obj.node as SearchableWithEdgeData).textSnippet;
+        return String(snippet ?? "");
+      },
+    }),
+  }),
+});
+
+builder.queryField("search", (t) =>
+  t.connection(
+    {
+      type: Searchable,
+      args: {
+        text: t.arg.string({ required: true }),
+      },
+      errors: {},
+      resolve: async (_, args, { session }) => {
+        return await resolveOffsetConnection(
+          { args },
+          async ({ limit, offset }) => {
+            const { text } = args;
+            return await prisma.$queryRaw<SearchableWithEdgeData[]>`
+        SELECT
+          "Searchable".*,
+          -- ranking function
+          ts_rank_cd(
+            setweight(to_tsvector(coalesce("Model".slug, '')), 'A') ||
+            setweight(to_tsvector(coalesce("SquiggleSnippet".code, '')), 'B') ||
+            setweight(to_tsvector(coalesce("RelativeValuesDefinition".slug, '')), 'A') ||
+            setweight(to_tsvector(coalesce("UserOwner".slug, '')), 'A') ||
+            setweight(to_tsvector(coalesce("GroupOwner".slug, '')), 'A'),
+            websearch_to_tsquery(${text})
+          ) AS rank,
+          ts_headline(
+            concat_ws(
+              ' ',
+              "Model".slug,
+              "RelativeValuesDefinition".slug,
+              "UserOwner".slug,
+              "GroupOwner".slug
+            ),
+            websearch_to_tsquery(${text}),
+            'HighlightAll=t'
+          ) AS "slugSnippet",
+          ts_headline(concat_ws(
+              ' ',
+              "SquiggleSnippet".code
+            ), websearch_to_tsquery(${text})
+          ) AS "textSnippet"
+        FROM "Searchable"
+        LEFT JOIN "Model" ON "Model".id = "Searchable"."modelId"
+          LEFT JOIN "Owner" AS "ModelOwner" ON "Model"."ownerId" = "ModelOwner".id
+            LEFT JOIN "User" AS "ModelOwnerUser" ON "ModelOwner".id = "ModelOwnerUser"."ownerId"
+            LEFT JOIN "Group" AS "ModelOwnerGroup" ON "ModelOwner".id = "ModelOwnerGroup"."ownerId"
+          LEFT JOIN "ModelRevision" ON "Model"."currentRevisionId" = "ModelRevision".id
+            LEFT JOIN "SquiggleSnippet" ON "ModelRevision"."contentId" = "SquiggleSnippet".id
+        LEFT JOIN "RelativeValuesDefinition" ON "RelativeValuesDefinition".id = "Searchable"."definitionId"
+          -- LEFT JOIN "Owner" AS "RelativeValuesDefinitionOwner" ON "RelativeValuesDefinition"."ownerId" = "RelativeValuesDefinitionOwner".id
+        LEFT JOIN "User" ON "User".id = "Searchable"."userId"
+          LEFT JOIN "Owner" AS "UserOwner" ON "User"."ownerId" = "UserOwner".id
+        LEFT JOIN "Group" ON "Group".id = "Searchable"."groupId"
+          LEFT JOIN "Owner" AS "GroupOwner" ON "Group"."ownerId" = "GroupOwner".id
```

Comment:
This kind of makes me want to denormalize everything in some batch process, like we'd need to do for algolia, but we can consider that later. Kudos for figuring this out.

Reply:
Yeah, I described an approximate roadmap in this PR's description.

> so we'll probably have to copy searchable fields (slugs, code, etc.) to JSON on Searchable table.